jenkins/Dockerfile

```diff
@@ -0,0 +1,45 @@
+FROM ubuntu:16.04
+
+RUN apt-get update
+RUN apt-get install -y curl software-properties-common git
+RUN curl -fsSL get.docker.com | sh
+
+##
+## Go
+##
+RUN add-apt-repository ppa:longsleep/golang-backports
+RUN apt-get update
+RUN apt-get install -y golang-go
+ENV GOPATH=$HOME/go
+ENV PATH=$PATH:$HOME/go/bin
+
+##
+## Go tools
+##
+RUN go get -u github.com/kardianos/govendor
+RUN go get -u github.com/jstemmer/go-junit-report
+RUN go get -u github.com/t-yuki/gocover-cobertura
+
+##
+## MongoDB
+##
+RUN apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 2930ADAE8CAF5059EE73BB4B58712A2291FA4AD5
+RUN echo "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu xenial/mongodb-org/3.6 multiverse" | \
+    tee /etc/apt/sources.list.d/mongodb-org-3.6.list
+RUN apt-get update
+RUN apt-get install -y mongodb-org
+
+##
+## Make jenkins can do anything
+##
+RUN apt-get install -y sudo
+RUN echo '%jenkins ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
+RUN chmod a+rw /etc/group
+RUN chmod a+rw /etc/passwd
+RUN echo 'jenkins:!::' >> /etc/gshadow
+RUN mkdir -p /home/jenkins && chmod -R a+rw /home/jenkins
```

Comment:
Jenkins is not installed in this image, how do we use Jenkins by only creating the directory?

Reply:
When Jenkins start docker container as a pipeline agent, $UID and $GID are passed into container. For example, Jenkins use `$UID:$GID` which is `113:117`
```
Jenkins does not seem to be running inside a container
$ docker run -t -d -u 113:117 --privileged --group-add docker ....
```

In the docker container, current user is not root any more but a unknown user with ID `113:117`.
And many things are denied since current user is not root. For example, `sh docker run ...` step is denied in the docker container.

To make Jenkinsfile can do anything without using docker args `-u 0:0`, I create a jenkins user inside the container and make `$UID:$GID` as same as the outside jenkins user.